### ISSUE TITLE
More efficient PISSA initialization for LoRA (faster/cheaper than current SVD)

### ISSUE THREAD

**Issue by Brokttv** [huggingface/peft#3049]
### Feature request

I'd like to propose (and potentially contribute) an improved version of the PiSSA initialization for LoRA adapters. I'm not sure if the current implementation uses `torch.svd_lowrank` (with optional fast subspace iteration via "pissa_niter_N"), which is already good but can still be quite expensive in time and/or memory for very large layers (e during adapter creation. So, if that the case, I would love to submit a PR.

CC: @BenjaminBossan 

### Your contribution

The algorithm mainly uses Krylov subspace method. I would share the code as soon as you can confirm if this was already harnessed or not.

**Comment by BenjaminBossan** [comment]
Thanks for your suggestion. PiSSA uses `svd` or `svd_lowrank`, depending on how it's configured:

https://github.com/huggingface/peft/blob/f6a7e678840a3e59c8e28f105695968f0dc706d4/src/peft/tuners/lora/layer.py#L316-L337

If you have a different algorithm (I haven't heard of Krylov subspace before), it could be added as another option if

1. It can be shown that it performs similarly or better (speed, memory, numerically)
2. Doesn't add external dependencies
3. SVD is still the default (backwards compatibility)

If this works, ideally as a drop in replacement for low rank SVD, that would be great, as we have other places in PEFT that could also benefit from that.

**Comment by Brokttv** [comment]
svd.low_rank's randomized SVD uses raw power iteration which causes numerical rank collapse.  My version adds QR re-orthogonalization at each step, giving 3–7 orders of magnitude better singular value accuracy at the same speed and nearly identical memory, all in pure PyTorch with no new dependencies.

Algorithm: 

[pissa.py](https://github.com/user-attachments/files/25400104/pissa.py)

Benchmark generated by Claude:

[pissa_benchmark.py](https://github.com/user-attachments/files/25400186/pissa_benchmark.py)

**Comment by BenjaminBossan** [comment]
Thanks for attaching the code and benchmark. I ran the benchmark on my machine (on a 4090) and Krylov SVD was not significantly better than low rank SVD. Did you get the same results?

```
======================================================================================================================
  rank=16  niter=2  dtype=torch.float32  device=cuda  seeds=5  timing_runs=10
======================================================================================================================
  Shape                      Method                     Time(ms)    vs PEFT    Mem(MB)   vs PEFT    FrobErr       SV-MSE
----------------------------------------------------------------------------------------------------------------------
  1k x 1k    (small)         svd_lowrank (PEFT)             0.78   (baseline)        4.0 (baseline)   1.74e-01     7.18e-05
                             Krylov SVD (proposed)          0.89     +13.5% x        4.4    +10.4%   1.70e-01     1.30e-07 v
----------------------------------------------------------------------------------------------------------------------
  2k x 2k    (medium)        svd_lowrank (PEFT)             0.91   (baseline)        3.6 (baseline)   1.81e-01     5.09e-05
                             Krylov SVD (proposed)          1.01     +11.9% x        4.5    +24.2%   1.77e-01     1.24e-07 v
----------------------------------------------------------------------------------------------------------------------
  4k x 4k    (Llama-7B attn) svd_lowrank (PEFT)             1.25   (baseline)        5.8 (baseline)   1.86e-01     4.14e-05
                             Krylov SVD (proposed)          1.31      +4.7% v        7.7    +32.8%   1.82e-01     1.37e-07 v
----------------------------------------------------------------------------------------------------------------------
  4k x 11k   (Llama-7B gate_proj) svd_lowrank (PEFT)        2.20   (baseline)        6.2 (baseline)   1.88e-01     5.57e-05
                             Krylov SVD (proposed)          2.34      +6.5% x       20.4   +229.1%   1.85e-01     2.55e-07 v
----------------------------------------------------------------------------------------------------------------------
  8k x 8k    (Llama-70B)     svd_lowrank (PEFT)             2.92   (baseline)       11.3 (baseline)   1.90e-01     7.71e-05
                             Krylov SVD (proposed)          3.12      +7.0% x       15.1    +33.4%   1.86e-01     1.51e-07 v
----------------------------------------------------------------------------------------------------------------------

======================================================================================================================
  rank=16  niter=4  dtype=torch.float32  device=cuda  seeds=5  timing_runs=10
======================================================================================================================
  Shape                      Method                     Time(ms)    vs PEFT    Mem(MB)   vs PEFT    FrobErr       SV-MSE
----------------------------------------------------------------------------------------------------------------------
  1k x 1k    (small)         svd_lowrank (PEFT)             1.08   (baseline)        4.0 (baseline)   1.71e-01     1.13e-05
                             Krylov SVD (proposed)          1.13      +4.9% v        4.4    +10.4%   1.70e-01     7.88e-09 v
----------------------------------------------------------------------------------------------------------------------
  2k x 2k    (medium)        svd_lowrank (PEFT)             1.28   (baseline)        3.6 (baseline)   1.78e-01     1.66e-05
                             Krylov SVD (proposed)          1.31      +2.8% v        4.5    +24.2%   1.77e-01     6.24e-09 v
----------------------------------------------------------------------------------------------------------------------
  4k x 4k    (Llama-7B attn) svd_lowrank (PEFT)             1.88   (baseline)        5.8 (baseline)   1.83e-01     1.14e-05
                             Krylov SVD (proposed)          1.78      -5.3% v        7.7    +32.8%   1.82e-01     1.31e-08 v
----------------------------------------------------------------------------------------------------------------------
  4k x 11k   (Llama-7B gate_proj) svd_lowrank (PEFT)        3.45   (baseline)        6.2 (baseline)   1.86e-01     1.34e-05
                             Krylov SVD (proposed)          3.39      -1.7% v       20.4   +229.1%   1.85e-01     1.75e-07 v
----------------------------------------------------------------------------------------------------------------------
  8k x 8k    (Llama-70B)     svd_lowrank (PEFT)             4.62   (baseline)       11.3 (baseline)   1.87e-01     1.15e-05
                             Krylov SVD (proposed)          4.69      +1.5% v       15.1    +33.4%   1.86e-01     2.75e-08 v
----------------------------------------------------------------------------------------------------------------------

======================================================================================================================
  rank=16  niter=8  dtype=torch.float32  device=cuda  seeds=5  timing_runs=10
======================================================================================================================
  Shape                      Method                     Time(ms)    vs PEFT    Mem(MB)   vs PEFT    FrobErr       SV-MSE
----------------------------------------------------------------------------------------------------------------------
  1k x 1k    (small)         svd_lowrank (PEFT)             1.71   (baseline)        4.0 (baseline)   1.70e-01     1.17e-06
                             Krylov SVD (proposed)          1.58      -7.6% v        4.4    +10.4%   1.70e-01     7.74e-09 v
----------------------------------------------------------------------------------------------------------------------
  2k x 2k    (medium)        svd_lowrank (PEFT)             2.07   (baseline)        3.6 (baseline)   1.77e-01     1.66e-06
                             Krylov SVD (proposed)          1.88      -9.2% v        4.5    +24.2%   1.77e-01     5.97e-09 v
----------------------------------------------------------------------------------------------------------------------
  4k x 4k    (Llama-7B attn) svd_lowrank (PEFT)             3.15   (baseline)        5.8 (baseline)   1.83e-01     3.20e-06
                             Krylov SVD (proposed)          2.68     -14.9% v        7.7    +32.8%   1.82e-01     1.31e-08 v
----------------------------------------------------------------------------------------------------------------------
  4k x 11k   (Llama-7B gate_proj) svd_lowrank (PEFT)        5.98   (baseline)        6.2 (baseline)   1.85e-01     1.95e-06
                             Krylov SVD (proposed)          5.45      -8.7% v       20.4   +229.1%   1.85e-01     1.75e-07 v
----------------------------------------------------------------------------------------------------------------------
  8k x 8k    (Llama-70B)     svd_lowrank (PEFT)             8.05   (baseline)       11.3 (baseline)   1.86e-01     1.81e-07
                             Krylov SVD (proposed)          7.82      -2.9% v       15.1    +33.4%   1.86e-01     2.75e-08 v
----------------------------------------------------------------------------------------------------------------------

```

Speed is roughly on par, memory is also similar or worse for Krylov. The error is lower, but I don't think that's a big issue in practice. So at least from this benchmark, the advantage is not really significant.

**Comment by Brokttv** [comment]
The premise that SVD accuracy won't be a big issue in practice directly contradicts the reason PiSSA was merged into PEFT in the first place, if init quality matters enough to prefer SVD over random, then it matters enough to prefer a more accurate SVD over a less accurate one.
 From the benchmark we have:
 1k x 1k:   svd_lowrank  7.18e-05  vs  Krylov  1.30e-07  →  550x worse
  4k x 4k:   svd_lowrank  4.14e-05  vs  Krylov  1.37e-07  →  300x worse
  8k x 8k:   svd_lowrank  7.71e-05  vs  Krylov  1.51e-07  →  500x worse

 svd_lowrank sits at e-05 and Krylov sits at e-07. That's a consistent 2-order-of-magnitude difference which is significant and not noise.

I would say a downstream fine-tuning experiment would be the definitive proof for you to see, but I don't have the compute for that right now. I'd argue the numerical evidence here is strong enough to warrant including this as an option.

**Comment by BenjaminBossan** [comment]
I focused on the advantages in terms of speed and memory, since this is what you described in your original post. Given that those didn't materialize, I wondered if there could be something wrong in the implementation. As you mentioned, in theory we should see advantages here.

When it comes to the error: It's true that Krylov SVD has a significantly lower error. For bigger models and more iterations, the advantage seems to shrink, but it could theoretically still matter. Of course, a certain level of imprecision is expected in this domain (e.g. loading a fp32 model in fp16 or bf16 incurs a similar MSE). At the end of the day, as you mention, we'd have to check a real example to be sure.

To proceed, how about creating a draft PR with a Krylov SVD function as drop-in replacement for low rank SVD (not PiSSA specific). Then I can run one of the included examples and check how it fares.

**Comment by Brokttv** [comment]
It's done.

**Comment by Brokttv** [comment]
Hi Benjamin, still waiting for your review.

**Comment by BenjaminBossan** [comment]
Closing for now, see discussion in #3056.
